plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,932 @@
+            __result = resultBuffer.detachOwn();
+
+            EVP_PKEY_CTX_free(encryptCtx);
+        }
+        catch (...)
+        {
+            if (encryptCtx)
+                EVP_PKEY_CTX_free(encryptCtx);
+            rtlFree(__result);
+            __lenResult = 0;
+            throw;
+        }
+    }
+}
+
+OPENSSL_API void OPENSSL_CALL rsaDecrypt(ICodeContext *ctx, size32_t & __lenResult, void * & __result, size32_t len_ciphertext, const void * _ciphertext, size32_t len_passphrase, const void * _passphrase, size32_t len_pem_private_key, const char * _pem_private_key)
+{
+    __result = nullptr;
+    __lenResult = 0;
+
+    // Initial sanity check of our arguments
+    if (len_pem_private_key == 0)
+        rtlFail(-1, "No private key provided");
+
+    EVP_PKEY_CTX * decryptCtx = nullptr;
+
+    if (len_ciphertext > 0)
+    {
+        try
+        {
+            // Load key from buffer
+            EVP_PKEY * privateKey = pkeyCache.checkCache(len_pem_private_key, _pem_private_key, len_passphrase, _passphrase);
+
+            // Create decryption context
+            decryptCtx = EVP_PKEY_CTX_new(privateKey, nullptr);
+            if (!decryptCtx)
+                failOpenSSLError("EVP_PKEY_CTX_new");
+            if (EVP_PKEY_decrypt_init(decryptCtx) <= 0)
+                failOpenSSLError("EVP_PKEY_decrypt_init");
+
+            // Determine max size of output
+            size_t outLen = 0;
+            if (EVP_PKEY_decrypt(decryptCtx, nullptr, &outLen, reinterpret_cast<const byte *>(_ciphertext), len_ciphertext) <= 0)
+                failOpenSSLError("EVP_PKEY_decrypt");
+
+            // Set actual size of output
+            MemoryBuffer resultBuffer(outLen);
+
+            if (EVP_PKEY_decrypt(decryptCtx, static_cast<byte *>(resultBuffer.bufferBase()), &outLen, reinterpret_cast<const byte *>(_ciphertext), len_ciphertext) <= 0)
+                failOpenSSLError("EVP_PKEY_decrypt");
+
+            __lenResult = outLen;
+            __result = resultBuffer.detachOwn();
+
+            EVP_PKEY_CTX_free(decryptCtx);
+        }
+        catch (...)
+        {
+            if (decryptCtx)
+                EVP_PKEY_CTX_free(decryptCtx);
+            rtlFree(__result);
+            __lenResult = 0;
+            throw;
+        }
+    }
+}
+
+OPENSSL_API void OPENSSL_CALL rsaSign(ICodeContext *ctx, size32_t & __lenResult, void * & __result, size32_t len_plaintext, const void * _plaintext, size32_t len_passphrase, const void * _passphrase, size32_t len_pem_private_key, const char * _pem_private_key, const char * _algorithm_name)
+{
+    EVP_MD_CTX *mdCtx = nullptr;
+
+    try
+    {
+        // Load the private key from the PEM string
+        EVP_PKEY * privateKey = pkeyCache.checkCache(len_pem_private_key, _pem_private_key, len_passphrase, _passphrase);
+
+        // Create and initialize the message digest context
+        mdCtx = EVP_MD_CTX_new();
+        if (!mdCtx)
+            failOpenSSLError("EVP_MD_CTX_new (rsaSign)");
```

Comment:
The 'rsaSign' is a misnomer, because you can sign with other public key types. This may need alteration elsewhere in the code as well.

Technically, there is a subset of public key types that can be used for signing. We can let OpenSSL generate an error message if necessary; no need to explicitly check. For reference:

_Only EVP_PKEY types that support signing can be used with these functions. This includes MAC algorithms where the MAC generation is considered as a form of "signing". Built-in EVP_PKEY types supported by these functions are CMAC, Poly1305, DSA, ECDSA, HMAC, RSA, SipHash, Ed25519 and Ed448._